[front] chore(stripe): Use SubscriptionResource in stripe webhooks

johnoppenheimer opened the pull request:
## Description
- Remove Subscription model direct calls inside stripe webhooks
- **Note**: `fetchActiveByWorkspace` will return a `FREE_NO_PLAN_SUBSCRIPTION_ID` type of subscription (which doesn't exists in DB), so need to use `isLegacyFreeNoPlan` to avoid updating a non existing subscription.
_To avoid mistakenly do that, we either could throw if `id < 0` in the many mutate method of the resource, or do the same directly in the `BaseResource`, open to suggestion_

## Tests
- [x] locally, signup to a new workspace and enter my stripe CC details, got subscribed correctly.
- [x] cancelling a subscription from Stripe move the workspace back to FREE_NO_PLAN

## Risk
- Mid, impact is on new / returning client, but only on their subscription, should have no impact on the daily usage of the product.

## Deploy Plan
- Deploy front

johnoppenheimer committed:
[front] chore(stripe): Use SubscriptionResource in stripe webhook

flvndvd commented on front/pages/api/stripe/webhook.ts:
```diff
@@ -309,10 +300,13 @@ async function handler(
-            where: { stripeSubscriptionId: invoice.subscription },
-            include: [Workspace],
-          });
+          // subscription = await Subscription.findOne({
```
✂️ ?

flvndvd commented on front/pages/api/stripe/webhook.ts:
```diff
@@ -352,10 +346,13 @@ async function handler(
-            where: { stripeSubscriptionId: invoice.subscription },
-            include: [Workspace],
-          });
+          // subscription = await Subscription.findOne({
```
✂️

flvndvd commented on front/pages/api/stripe/webhook.ts:
```diff
@@ -377,12 +374,21 @@ async function handler(
+            await subscription.updatePaymentFailingSince(now);
+          }
+
+          const workspace = await Workspace.findByPk(subscription.workspaceId);
```
Let's make sure we use a function from lib/api/workspace.ts. IIRC we have something like `unsafeGetWorkspacesByModelId`.

flvndvd commented on front/pages/api/stripe/webhook.ts:
```diff
@@ -464,10 +470,13 @@ async function handler(
+              const subscription = await SubscriptionResource.fetchByStripeId(
+                stripeSubscription.id
+              );
+              // const subscription = await Subscription.findOne({
```
✂️

johnoppenheimer committed:
make method to unsafe get model workspace by model id

johnoppenheimer committed:
fix types

johnoppenheimer committed:
use cleaner function

johnoppenheimer committed:
use WorkspaceResource in stripe webhook

johnoppenheimer committed:
throw error if workspace associated to a sub is not found

also create a authenticator helper to avoid doing the same requests
multiple time

johnoppenheimer committed:
verify if its a dummy subscription before trying to end it
